File: 0046-permutations/0046-permutations.cs
public class Solution {
    public IList<IList<int>> Permute(int[] nums) {

        IList<IList<int>> ans=new List<IList<int>>();
        bool[] arr=new bool[nums.Length];
        check(nums,arr,ans,new List<int>());
        return ans;
    }

    public void check(int[] nums,bool[] arr,IList<IList<int>> ans,IList<int> demo){
        if(demo.Count==nums.Length){
            ans.Add(new List<int>(demo));
            return;
        }

        for(int i=0;i<nums.Length;i++){
            if(!arr[i]){
                demo.Add(nums[i]);
                arr[i]=true;
                check(nums,arr,ans,demo);
                demo.RemoveAt(demo.Count-1);
                arr[i]=false;
            }
        }
    }
}
    

File: 0046-permutations/0046-permutations.java
class Solution {
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> ans=new ArrayList<List<Integer>>();
        boolean arr[]=new boolean[nums.length];
        check(nums,arr,ans,new ArrayList<>());
        return ans;
    }

    public void check(int[] nums,boolean arr[],List<List<Integer>> ans,ArrayList<Integer> demo){
        if(demo.size()==nums.length){
            ans.add(new ArrayList<>(demo));
            return;
        }

        for(int i=0;i<nums.length;i++){
            if(!arr[i]){
                demo.add(nums[i]);
                arr[i]=true;
                check(nums,arr,ans,demo);
                demo.remove(demo.size()-1);
                arr[i]=false;
            }
        }
    }
}